rfcs/014-pagination.md

```diff
@@ -0,0 +1,193 @@
+- add `page_size` query field to control number of rows on each page and to flag
+  that client expects paginated response
+- add `first`, `previous`, `next` fields which contain bookmark part of URI
+- add `bookmark` query field to retrieve bookmarked page
+
+
+## Requirements Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+- `bookmark` - is opaque token which would contain information needed to retrieve
+  bookmarked page. The format of the token value MUST NOT be relied upon in the client.
+
+---
+
+# Detailed Description
+
+## Implementation proposal
+
+1) Add new optional query field called `bookmark` to following endpoints
+  - {db}/_all_docs
+  - {db}/_all_docs/queries
+  - _dbs_info
+  - {db}/_design/{ddoc}/_view/{view}
+  - {db}/_design/{ddoc}/_view/{view}/queries
+
+2) Add new optional query field called `page_size` if it is set we would use
+  paginated endpoint otherwise use old code path
+3) Implement per-endpoint configurable max limits to restrict page size of
+  paginated responses
+
+ ```
+ [request_limits]
+ _all_docs = 5000
+ _all_docs/queries = 5000
+ _all_dbs = 5000
+ _dbs_info = 5000
+ _view = 2500
+ _view/queries = 2500
+ _find = 2500
+ ```
+4) Add following additional fields into response:
+  ```
+    "first": "12345678945621321689",
+    "previous": "983uiwfjkdsdf",
+    "next": "12343tyekf3"
+  ```
+
+## Limitations
+
+- The `first`/`next`/`last` keys in the response are represented as path which
+  includes the bookmark query key. This means the bookmark token size contributes
+  to total URI length and is subject to a max URL lenght (8192 these days).
+  CouchDB historically didn't have limit on docid. FDB imposes a limit on a key path.
+  Therefore we would probably have a strict limit in CouchDB 4.0. We didn't agree
+  on the maximum length of docid yet. However it seems to be trending towards 5kB
+  (see [RFC 008's comments](https://github.com/apache/couchdb-documentation/pull/410#discussion_r340536640)).
+  With 5kB sized keys we would exceed 8192 characters when both `start_key` and
+  `end_key` would be specified. This means we would have to support POST requests.
+
+## Semantics of the implementation
+
+- The bookmark would include information needed to ensure proper pagination
+  without the need to repeat initial parameters of the request.
+- Don't use delayed responses when `bookmark` field is provided
+- Don't use delayed responses when `page_size` query key is specified and when
+  it is below the max limit
+- Return 400 when `bookmark` field is set and other query fields are present
+- Return 400 when `page_size` query key is specified and it is greater than
+  the max limit
+- Return 400 when we stream rows (`page_size` query key wasn't specified) and
+  reach max limit
```

Comment:
How would this work with streaming? If we stream the request it implies we already emitted the 200 OK response. I think we'd have to accumulate the results here and only emit them after we know they are less than the limit or emit an a 400 if there are too many?

Reply:
Good question!